File: cell/liveCell_test.go
package cell

import (
	"testing"

	assert "github.com/stretchr/testify/assert"
)

func TestNewLiveCell(t *testing.T) {

	t.Run("Live cell is created.", func(t *testing.T) {
		got := NewLiveCell()
		// assert true for IsAlive

		assert.True(t, got.IsAlive())
	})
}

func TestNextGenerationOfLiveCell(t *testing.T) {
	type args struct {
		neighbours int
	}

	type test struct {
		name     string
		args     args
		want     Cell
		expected error
	}

	tests := []test{
		{
			name:     "Live cell with 2 alive neighbours stays alive.",
			args:     args{neighbours: 2},
			want:     NewLiveCell(),
			expected: nil,
		},
		{
			name:     "Live cell with 3 alive neighbours stays alive.",
			args:     args{neighbours: 3},
			want:     NewLiveCell(),
			expected: nil,
		},
		{
			name:     "Live cell with less than 2 alive neighbours dies.",
			args:     args{neighbours: 1},
			want:     NewDeadCell(),
			expected: nil,
		},
		{
			name:     "Live cell with more than 3 alive neighbours dies.",
			args:     args{neighbours: 4},
			want:     NewDeadCell(),
			expected: nil,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			got, err := NewLiveCell().NextGeneration(tt.args.neighbours)
			assert.Equal(t, tt.expected, err)
			assert.Equal(t, tt.want, got)
		})
	}
}

File: cell/liveCell.test.js
const { DeadCell, LiveCell } = require("./cell");

describe("LiveCell", () => {
  let liveCell;

  beforeEach(() => {
    liveCell = new LiveCell();
  });

  it("should return true", () => {
    expect(liveCell.isAlive()).toBe(true);
  });

  describe("nextGeneration", () => {
    it("should return a DeadCell if it has less than 2 live neighbours", () => {
      expect(liveCell.nextGeneration(0) instanceof DeadCell).toBe(true);
      expect(liveCell.nextGeneration(1) instanceof DeadCell).toBe(true);
    });

    it("should return a LiveCell if it has 2 or 3 live neighbours", () => {
      expect(liveCell.nextGeneration(2) instanceof LiveCell).toBe(true);
      expect(liveCell.nextGeneration(3) instanceof LiveCell).toBe(true);
    });

    it("should return a DeadCell if it has more than 3 live neighbours", () => {
      expect(liveCell.nextGeneration(4) instanceof DeadCell).toBe(true);
    });
  });
});
